packages/central-server/app/patientPortal/auth/PortalOneTimeTokenService.js

```diff
@@ -0,0 +1,114 @@
+import { addMinutes } from 'date-fns';
+import { randomInt, randomBytes } from 'crypto';
+import bcrypt from 'bcrypt';
+import { PORTAL_ONE_TIME_TOKEN_TYPES } from '@tamanu/constants';
+import { BadAuthenticationError } from '@tamanu/shared/errors';
+
+const DEFAULT_SALT_ROUNDS = 10;
+
+function randomSixDigitCode() {
+  // returns a zero-padded 6 digit string using crypto.randomInt for better security
+  return Array.from({ length: 6 }, () => randomInt(0, 9)).join('');
+}
+
+function randomRegisterCode() {
+  // secure random code for magic links - 32 hex chars
+  return randomBytes(16).toString('hex');
+}
+
+export async function hashPortalToken(token) {
+  return bcrypt.hash(token, DEFAULT_SALT_ROUNDS);
+}
+
+export class PortalOneTimeTokenService {
+  constructor(models, { expiryMinutes = 10 } = {}) {
+    this.models = models;
+    this.expiryMinutes = expiryMinutes;
+  }
+
+  async createLoginToken(portalUserId) {
+    const { PortalOneTimeToken } = this.models;
+    const token = randomSixDigitCode();
+    const hashedToken = await hashPortalToken(token);
+    const expiresAt = addMinutes(new Date(), this.expiryMinutes);
+
+    // Overwrite existing login tokens for this user
+    await PortalOneTimeToken.destroy({
+      where: { portalUserId, type: PORTAL_ONE_TIME_TOKEN_TYPES.LOGIN },
+      force: true,
+    });
+
+    const record = await PortalOneTimeToken.create({
+      portalUserId,
+      type: PORTAL_ONE_TIME_TOKEN_TYPES.LOGIN,
+      token: hashedToken,
+      expiresAt,
+    });
+    return {
+      token,
+      expiresAt: record.expiresAt,
+    };
+  }
+
+  async createRegisterToken(portalUserId) {
+    const { PortalOneTimeToken } = this.models;
+    const token = randomRegisterCode();
+    const hashedToken = await hashPortalToken(token);
+    const expiresAt = addMinutes(new Date(), this.expiryMinutes);
+
+    // Overwrite existing register tokens for this user
+    await PortalOneTimeToken.destroy({
+      where: { portalUserId, type: PORTAL_ONE_TIME_TOKEN_TYPES.REGISTER },
+      force: true,
+    });
+
+    const record = await PortalOneTimeToken.create({
+      portalUserId,
+      type: PORTAL_ONE_TIME_TOKEN_TYPES.REGISTER,
+      token: hashedToken,
+      expiresAt,
+    });
+    return {
+      token,
+      expiresAt: record.expiresAt,
+    };
+  }
+
+  async verifyAndConsume({ token, type = PORTAL_ONE_TIME_TOKEN_TYPES.LOGIN }) {
+    const { PortalOneTimeToken, PortalUser } = this.models;
+
+    // Find all records of this type to compare against
+    const records = await PortalOneTimeToken.findAll({
+      where: { type },
+    });
+
+    let matchingRecord = null;
+
+    // Compare the token against all stored hashes
+    for (const record of records) {
```

Comment:
I'm not sure how intensive this will be to bcrypt.compare in a loop? I'm not sure how to get around it as we don't know. the user when verifying these codes and links. Noting that we do destroy used tokens which helps.

I initially added a filter to remove expired tokens which would ensure the list of tokens is small but then we wouldn't be able to feedback to the user if the token is expired.

Reply:
have updated the verification to require a portal user id so that we no longer have to bcrypt.compare multiple times